README
NAME
    indirect - Lexically warn about using the indirect object syntax.

VERSION
    Version 0.14

SYNOPSIS
        # In a script
        no indirect;
        my $x = new Apple 1, 2, 3; # warns
        {
         use indirect;
         my $y = new Pear; # ok
         {
          no indirect hook => sub { die "You really wanted $_[0]\->$_[1]" };
          my $z = new Pineapple 'fresh'; # croaks 'You really wanted Pineapple->new'
         }
        }
        no indirect ':fatal';
        if (defied $foo) { ... } # croaks, note the typo

        # From the command-line
        perl -M-indirect -e 'my $x = new Banana;' # warns

        # Or each time perl is ran
        export PERL5OPT="-M-indirect"
        perl -e 'my $y = new Coconut;' # warns

DESCRIPTION
    When enabled (or disabled as some may prefer to say, since you actually
    turn it on by calling "no indirect"), this pragma warns about indirect
    object syntax constructs that may have slipped into your code. This
    syntax is now considered harmful, since its parsing has many quirks and
    its use is error prone (when "swoosh" isn't defined, "swoosh $x"
    actually compiles to "$x->swoosh").

    It currently does not warn when the object is enclosed between braces
    (like "meth { $obj } @args") or for core functions ("print" or "say").
    This may change in the future, or may be added as optional features that
    would be enabled by passing options to "unimport".

    This module is not a source filter.

METHODS
  "unimport [ hook => $hook | ':fatal' ]"
    Magically called when "no indirect @opts" is encountered. Turns the
    module on. The policy to apply depends on what is first found in @opts :

    *   If it's the string ':fatal', the compilation will croak on the first
        indirect syntax met.

    *   If the key/value pair "hook => $hook" comes first, $hook will be
        called for each error with the object name as $_[0], the method name
        as $_[1], the current file as $_[2] and the line number as $_[3].

    *   Otherwise, a warning will be emitted for each indirect construct.

  "import"
    Magically called at each "use indirect". Turns the module off.

CONSTANTS
  "I_THREADSAFE"
    True iff the module could have been built when thread-safety features.

CAVEATS
    The implementation was tweaked to work around several limitations of
    vanilla "perl" pragmas : it's thread safe, and doesn't suffer from a
    "perl 5.8.x-5.10.0" bug that causes all pragmas to propagate into
    "require"d scopes.

    "meth $obj" (no semicolon) at the end of a file won't be seen as an
    indirect object syntax, although it will as soon as there is another
    token before the end (as in "meth $obj;" or "meth $obj 1").

    With 5.8 perls, the pragma does not propagate into "eval STRING". This
    is due to a shortcoming in the way perl handles the hints hash, which is
    addressed in perl 5.10.

DEPENDENCIES
    perl 5.8.

    XSLoader (standard since perl 5.006).

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

    You can contact me by mail or on "irc.perl.org" (vincent).

BUGS
    Please report any bugs or feature requests to "bug-indirect at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=indirect>. I will be
    notified, and then you'll automatically be notified of progress on your
    bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc indirect

    Tests code coverage report is available at
    <http://www.profvince.com/perl/cover/indirect>.

ACKNOWLEDGEMENTS
    Bram, for motivation and advices.

COPYRIGHT & LICENSE
    Copyright 2008-2009 Vincent Pit, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for indirect

0.14    2009-06-04 21:55 UTC
        + Fix : Prevent bogus invalid syntaxes caused by reallocated memory
                chunks. Thanks Andrew Main for reporting with a reproducible
                test case.

0.13    2009-05-24 18:50 UTC
        + Add : The hook now receives the file name and the line where the
                error happened in respectively $_[2] and $_[3].
        + Fix : Pass mortalized copies of the method name and object to the
                hook. This seems to fix some rare crashes.
        + Fix : Work around a bug in perl 5.10.0 and lower. Thanks Andrew Main
                for teaching me about this issue.
        + Fix : Report the correct file in error messages (a regression from the
                previous version).

0.12    2009-05-03 14:30 UTC
        + Add : You can specify the action to execute for each indirect
                construct encountered with the new "hook => $coderef" unimport()
                option.
        + Chg : A ptable is now used internally for the op => position mapping.
        + Fix : The pragma is now built with thread safety features enabled
                whenever possible (a notable exception is perl 5.8.x on Win32,
                as something seems wrong with its context handling).
                The new indirect::I_THREADSAFE() constant reflects this.
        + Fix : A negation precedence nit in indirect_ck_entersub().
        + Tst : "use/no indirect" while parsing an indirect construct.
        + Tst : Thread safety.

0.11    2009-02-08 18:35 UTC
        + Fix : Potential collisions by hashing pointers with a wrong format.
        + Upd : Resources in META.yml.

0.10    2009-01-17 12:40 UTC
        Re-release 0.09_01 as stable.

0.09_01 2008-12-08 17:55 UTC
        + Fix : Invalid memory read with "${\(new Foo)}" constructs. The
                corresponding test is turned back on.
        + Tst : Refinements in t/30-scope.t

0.09    2008-12-05 20:35 UTC
        + Add : Support for perl 5.8.
        + Tst : Skip a test in t/10-good.t that randomly segfaults for (I guess)
                systems stricter than linux in the way they manage their memory.

0.08    2008-10-22 14:45 UTC
        + Fix : A rare edge case for package whose names are prefix of 'main'.
        + Tst : Test $$ as variable and state variables.

0.07_03 2008-10-17 20:10 UTC
        + Add : Support and tests for variables with spaces after the sigil.
        + Upd : META.yml spec updated to 1.4.

0.07_02 2008-10-15 21:10 UTC
        + Add : Support and tests for package variables.
        + Tst : Coverage improved by removing dead code.

0.07_01 2008-10-15 16:00 UTC
        + Fix : [RT #40055] : Not handling RV2SV => GV(SV) correctly, which
                could cause 'no indirect; print' segfaults. Thanks Goro Fuji for
                reporting.

0.06    2008-10-11 16:45 UTC
        + Doc : Nits.
        + Tst : Test "no indirect 'anything'", "foo Class->bar", and indirect
                uses of exec() and system().

0.05    2008-10-02 14:40 UTC
        + Chg : Now the old check function is always called before storing an op
                into the map.
        + Fix : Misc code and docs refinements.

0.04    2008-08-30 19:00 UTC
        + Fix : Clean up the op->src hash when we're done with an entersub.
        + Tst : No longer fork for testing. IPC::Cmd isn't required anymore.

0.03    2008-08-12 15:25 UTC
        This release is kindly supported by Copenhagen Hotel Centrum WiFi.
        + Fix : Tests used not to pass PERL5OPTS to their kids. This lead to
                failures under CPAN. I think.
        + Tst : Refinements.

0.02    2008-08-11 15:55 UTC
        + Fix : Some building failures with old gcc versions that didn't seem to
                like the ((hint == 2) ? croak : warn)(msg) construct. I think.
        + Rem : Unused cruft from a previous implementation.
        + Tst : Fail more gracefully when we can't capture buffers or when the
                child returned an error.

0.01    2008-08-10 20:40 UTC
        First version, released on an unsuspecting world.



t/data/bad.d
#!perl

use strict;
use warnings;

my $obj;
my $pkg;
my $cb;

$obj = new Hlagh1;
$obj = new Hlagh2();
$obj = new Hlagh3(1);
$obj = new Hlagh4(1, 2);

$obj = new        Hlagh5            ;
$obj = new        Hlagh6     (      )      ;
$obj = new        Hlagh7     (      1        )     ;
$obj = new        Hlagh8     (      1        ,       2        )     ;

$obj = new    
                      Hlagh9		
        ;
$obj = new   
                                       Hlagh10     (    
                  )      ;
$obj =
              new    
    Hlagh11     (      1   
            )     ;
$obj =
new      
Hlagh12    
                   (      1        ,  
                2        )     ;

my $x;
$obj = new13 $x;
$obj = new14 $x();
$obj = new15 $x('foo');
$obj = new16 $x qq{foo}, 1;
$obj = new17 $x qr{foo\s+bar}, 1 .. 1;
$obj = new18 $x(qw/bar baz/);

$obj = new19
          $_;
$obj = new20
             $_     (        );
$obj = new21 $_      qr/foo/  ;
$obj = new22 $_     qq(bar baz);

potato23 $x;
potato24 $x, 1, 2;

$obj = Hlagh25Hlagh25 Hlagh25;
$obj = Hlagh26Hlagh26 Hlagh26; # Hlagh26Hlagh26 Hlagh26
$obj = new27 new27new27;
$obj = new28 new28new28; # new28 new28new28


t/data/good.d
#!perl

use strict;
use warnings;

my $obj;
my $pkg;
my $cb;

$obj = Hlagh1->new;
$obj = Hlagh2->new();
$obj = Hlagh3->new(1);
$obj = Hlagh4->new(q{foo}, bar => $obj);

$obj = Hlagh5   ->   new   ;
$obj = Hlagh6   ->   new   (   )   ;
$obj = Hlagh7   ->   new   (   1   )   ;
$obj = Hlagh8   ->   new   (   'foo'   ,   bar =>   $obj   );

$obj = Hlagh9
            ->
                          new   ;
$obj = Hlagh10  

      ->   
new   ( 
 )   ;
$obj = Hlagh11
                                       ->   new   ( 
               1   )   ;
$obj = Hlagh12   ->
                              new   (   "foo"
  ,    bar     
               =>        $obj       );

$obj = Hlagh13->$cb;
$obj = Hlagh14->$cb();
$obj = Hlagh15->$cb($pkg);
$obj = Hlagh16->$cb(sub { 'foo' },  bar => $obj);

$obj = $pkg->new17   ;
$obj = $pkg  ->   new18  (   );
$obj = $pkg       
           -> 
        new19 ( $pkg );
$obj = 
         $pkg
->
new20        (     qr/foo/,
      foo => qr/bar/   );

$obj 
  =  
$pkg
->
$cb
;
$obj = $pkg    ->   ($cb)   ();
$obj = $pkg->$cb( $obj  );
$obj = $pkg->$cb(qw/foo bar baz/);

my $x;

$obj = new25 { $x };
$obj = new26
  {
     $x  }
  ();
$obj = new27 {
  $x  } qq/foo/;
$obj = new28
   {
      $x
    }(qw/bar baz/);

sub potato29;
sub potato30;

potato29 $x;
potato30 $x, 1 , 2;

print STDOUT "bananananananana\n";
print $x "oh hai\n";


t/data/mixed.d
#!perl

use strict;
use warnings;

my $a = new P1;

{
 no indirect;
 my $b = new P2;
 {
  my $c = new P3;
 }
 {
  use indirect;
  my $d = new P4;
 }
 my $e = new P5;
}

my $f = new P6;

no indirect;

my $g = new P7;

use indirect;

my $h = new P8;


ptable.h
/* This file is part of the indirect Perl module.
 * See http://search.cpan.org/dist/indirect/ */

/* This is a pointer table implementation essentially copied from the ptr_table
 * implementation in perl's sv.c, except that it has been modified to use memory
 * shared across threads.
 * Copyright goes to the original authors, bug reports to me. */

/* This header is designed to be included several times with different
 * definitions for PTABLE_NAME and PTABLE_VAL_FREE(). */

#undef pPTBLMS
#undef pPTBLMS_
#undef aPTBLMS
#undef aPTBLMS_

/* Context for PerlMemShared_* functions */

#ifdef PERL_IMPLICIT_SYS
# define pPTBLMS  pTHX
# define pPTBLMS_ pTHX_
# define aPTBLMS  aTHX
# define aPTBLMS_ aTHX_
#else
# define pPTBLMS
# define pPTBLMS_
# define aPTBLMS
# define aPTBLMS_
#endif

#ifndef pPTBL
# define pPTBL  pPTBLMS
#endif
#ifndef pPTBL_
# define pPTBL_ pPTBLMS_
#endif
#ifndef aPTBL
# define aPTBL  aPTBLMS
#endif
#ifndef aPTBL_
# define aPTBL_ aPTBLMS_
#endif

#ifndef PTABLE_NAME
# define PTABLE_NAME ptable
#endif

#ifndef PTABLE_VAL_FREE
# define PTABLE_VAL_FREE(V)
#endif

#ifndef PTABLE_JOIN
# define PTABLE_PASTE(A, B) A ## B
# define PTABLE_JOIN(A, B)  PTABLE_PASTE(A, B)
#endif

#ifndef PTABLE_PREFIX
# define PTABLE_PREFIX(X) PTABLE_JOIN(PTABLE_NAME, X)
#endif

#ifndef ptable_ent
typedef struct ptable_ent {
 struct ptable_ent *next;
 const void *       key;
 void *             val;
} ptable_ent;
#define ptable_ent ptable_ent
#endif /* !ptable_ent */

#ifndef ptable
typedef struct ptable {
 ptable_ent **ary;
 UV           max;
 UV           items;
} ptable;
#define ptable ptable
#endif /* !ptable */

#ifndef ptable_new
STATIC ptable *ptable_new(pPTBLMS) {
#define ptable_new() ptable_new(aPTBLMS)
 ptable *t = PerlMemShared_malloc(sizeof *t);
 t->max   = 15;
 t->items = 0;
 t->ary   = PerlMemShared_calloc(t->max + 1, sizeof *t->ary);
 return t;
}
#endif /* !ptable_new */

#ifndef PTABLE_HASH
# define PTABLE_HASH(ptr) \
     ((PTR2UV(ptr) >> 3) ^ (PTR2UV(ptr) >> (3 + 7)) ^ (PTR2UV(ptr) >> (3 + 17)))
#endif

#ifndef ptable_find
STATIC ptable_ent *ptable_find(const ptable * const t, const void * const key) {
#define ptable_find ptable_find
 ptable_ent *ent;
 const UV hash = PTABLE_HASH(key);

 ent = t->ary[hash & t->max];
 for (; ent; ent = ent->next) {
  if (ent->key == key)
   return ent;
 }

 return NULL;
}
#endif /* !ptable_find */

#ifndef ptable_fetch
STATIC void *ptable_fetch(const ptable * const t, const void * const key) {
#define ptable_fetch ptable_fetch
 const ptable_ent *const ent = ptable_find(t, key);

 return ent ? ent->val : NULL;
}
#endif /* !ptable_fetch */

#ifndef ptable_split
STATIC void ptable_split(pPTBLMS_ ptable * const t) {
#define ptable_split(T) ptable_split(aPTBLMS_ (T))
 ptable_ent **ary = t->ary;
 const UV oldsize = t->max + 1;
 UV newsize = oldsize * 2;
 UV i;

 ary = PerlMemShared_realloc(ary, newsize * sizeof(*ary));
 Zero(&ary[oldsize], newsize - oldsize, sizeof(*ary));
 t->max = --newsize;
 t->ary = ary;

 for (i = 0; i < oldsize; i++, ary++) {
  ptable_ent **curentp, **entp, *ent;
  if (!*ary)
   continue;
  curentp = ary + oldsize;
  for (entp = ary, ent = *ary; ent; ent = *entp) {
   if ((newsize & PTABLE_HASH(ent->key)) != i) {
    *entp     = ent->next;
    ent->next = *curentp;
    *curentp  = ent;
    continue;
   } else
    entp = &ent->next;
  }
 }
}
#endif /* !ptable_split */

STATIC void PTABLE_PREFIX(_store)(pPTBL_ ptable * const t, const void * const key, void * const val) {
 ptable_ent *ent = ptable_find(t, key);

 if (ent) {
  void *oldval = ent->val;
  PTABLE_VAL_FREE(oldval);
  ent->val = val;
 } else if (val) {
  const UV i = PTABLE_HASH(key) & t->max;
  ent = PerlMemShared_malloc(sizeof *ent);
  ent->key  = key;
  ent->val  = val;
  ent->next = t->ary[i];
  t->ary[i] = ent;
  t->items++;
  if (ent->next && t->items > t->max)
   ptable_split(t);
 }
}

#ifndef ptable_walk
STATIC void ptable_walk(pTHX_ ptable * const t, void (*cb)(pTHX_ ptable_ent *ent, void *userdata), void *userdata) {
#define ptable_walk(T, CB, UD) ptable_walk(aTHX_ (T), (CB), (UD))
 if (t && t->items) {
  register ptable_ent ** const array = t->ary;
  UV i = t->max;
  do {
   ptable_ent *entry;
   for (entry = array[i]; entry; entry = entry->next)
    cb(aTHX_ entry, userdata);
  } while (i--);
 }
}
#endif /* !ptable_walk */

STATIC void PTABLE_PREFIX(_clear)(pPTBL_ ptable * const t) {
 if (t && t->items) {
  register ptable_ent ** const array = t->ary;
  UV i = t->max;

  do {
   ptable_ent *entry = array[i];
   while (entry) {
    ptable_ent * const oentry = entry;
    void *val = oentry->val;
    entry = entry->next;
    PTABLE_VAL_FREE(val);
    PerlMemShared_free(oentry);
   }
   array[i] = NULL;
  } while (i--);

  t->items = 0;
 }
}

STATIC void PTABLE_PREFIX(_free)(pPTBL_ ptable * const t) {
 if (!t)
  return;
 PTABLE_PREFIX(_clear)(aPTBL_ t);
 PerlMemShared_free(t->ary);
 PerlMemShared_free(t);
}

#undef pPTBL
#undef pPTBL_
#undef aPTBL
#undef aPTBL_

#undef PTABLE_NAME
#undef PTABLE_VAL_FREE


Makefile.PL
use 5.008;

use strict;
use warnings;
use ExtUtils::MakeMaker;

my @DEFINES;

# Threads, Windows and 5.8.x don't seem to be best friends
if ($^O eq 'MSWin32' && $^V lt v5.9.0) {
 push @DEFINES, '-DI_MULTIPLICITY=0';
}

@DEFINES = (DEFINE => join ' ', @DEFINES) if @DEFINES;

my $dist = 'indirect';

my %META = (
 configure_requires => {
  'ExtUtils::MakeMaker' => 0,
 },
 build_requires => {
  'ExtUtils::MakeMaker' => 0,
  'Test::More'          => 0,
 },
 resources => {
  bugtracker => "http://rt.cpan.org/NoAuth/ReportBug.html?Queue=$dist",
  homepage   => "http://search.cpan.org/dist/$dist/",
  license    => 'http://dev.perl.org/licenses/',
  repository => "http://git.profvince.com/?p=perl%2Fmodules%2F$dist.git",
 },
);

WriteMakefile(
    NAME             => 'indirect',
    AUTHOR           => 'Vincent Pit <perl@profvince.com>',
    LICENSE          => 'perl',
    VERSION_FROM     => 'lib/indirect.pm',
    ABSTRACT_FROM    => 'lib/indirect.pm',
    PL_FILES         => {},
    @DEFINES,
    PREREQ_PM        => {
        'XSLoader' => 0,
    },
    MIN_PERL_VERSION => 5.008,
    META_MERGE       => \%META,
    dist             => {
        PREOP    => 'pod2text lib/indirect.pm > $(DISTVNAME)/README',
        COMPRESS => 'gzip -9f', SUFFIX => 'gz'
    },
    clean            => {
        FILES => "$dist-* *.gcov *.gcda *.gcno cover_db Debian_CPANTS.txt"
    },
);


samples/indirect.pl
#!/usr/bin/env perl

#use strict;
#use warnings;

use lib qw{blib/lib blib/arch};

sub Hlagh::new { my $class = shift; bless { }, ref($class) || $class ; }

sub foo { shift; print "foo $_[0]\n" }
sub bar { print "wut\n"; }
my $bar = bless { }, 'main';

my %h;
my $x = 1;

no indirect;

$x = new Hlagh 1, 2, 3;
my $y = slap $x "what", 5;
$h{foo} = 12;

use indirect;

foo 4, 5;

no indirect;

my $pkg = 'Hlagh';
my $cb = 'new';

foo(6, 7, 8); my $y = new $_ qr/bar/;

my $y = Hlagh->new;
$y =          new Hlagh;
my $z = foo meh, 1, 2;
$y = meh $x, 7;
$y = foo(3, 4);
$y = Hlagh->new();
$y = Hlagh->new(1, 2, 3);
$y = Hlagh->$cb;
$y =      new Hlagh;
$y =                new Hlagh 1, 2, 3;
$y = 
  new   
     Hlagh
        1  ,
                2,        3;
$y = new $pkg;
$y = new $pkg 'what';
$y = $pkg->new;
$y = $pkg->new(1, 2, 3);
$y = $pkg->$cb;
$y = new(Hlagh);
$y = new { Hlagh };
$y = new { $y };
$y = Hlagh
        ->        new 
           ( 1     ,       2,    3);
$y = Hlagh
        ->        $ cb
           ( 1     ,       2,    3);
$y = new Hlagh $,;
$y = new Hlagh ',';
print { $^H{dongs} } 'bleh';
print STDERR 1;
print STDERR 'what';
print STDERR q{wat};
my $fh;
print $fh 'dongs';


lib/indirect.pm
package indirect;

use 5.008;

use strict;
use warnings;

=head1 NAME

indirect - Lexically warn about using the indirect object syntax.

=head1 VERSION

Version 0.14

=cut

our $VERSION;
BEGIN {
 $VERSION = '0.14';
}

=head1 SYNOPSIS

    # In a script
    no indirect;
    my $x = new Apple 1, 2, 3; # warns
    {
     use indirect;
     my $y = new Pear; # ok
     {
      no indirect hook => sub { die "You really wanted $_[0]\->$_[1]" };
      my $z = new Pineapple 'fresh'; # croaks 'You really wanted Pineapple->new'
     }
    }
    no indirect ':fatal';
    if (defied $foo) { ... } # croaks, note the typo

    # From the command-line
    perl -M-indirect -e 'my $x = new Banana;' # warns

    # Or each time perl is ran
    export PERL5OPT="-M-indirect"
    perl -e 'my $y = new Coconut;' # warns

=head1 DESCRIPTION

When enabled (or disabled as some may prefer to say, since you actually turn it on by calling C<no indirect>), this pragma warns about indirect object syntax constructs that may have slipped into your code.
This syntax is now considered harmful, since its parsing has many quirks and its use is error prone (when C<swoosh> isn't defined, C<swoosh $x> actually compiles to C<< $x->swoosh >>).

It currently does not warn when the object is enclosed between braces (like C<meth { $obj } @args>) or for core functions (C<print> or C<say>).
This may change in the future, or may be added as optional features that would be enabled by passing options to C<unimport>.

This module is B<not> a source filter.

=cut

BEGIN {
 require XSLoader;
 XSLoader::load(__PACKAGE__, $VERSION);
}

=head1 METHODS

=head2 C<< unimport [ hook => $hook | ':fatal' ] >>

Magically called when C<no indirect @opts> is encountered.
Turns the module on.
The policy to apply depends on what is first found in C<@opts> :

=over 4

=item *

If it's the string C<':fatal'>, the compilation will croak on the first indirect syntax met.

=item *

If the key/value pair C<< hook => $hook >> comes first, C<$hook> will be called for each error with the object name as C<$_[0]>, the method name as C<$_[1]>, the current file as C<$_[2]> and the line number as C<$_[3]>.

=item *

Otherwise, a warning will be emitted for each indirect construct.

=back

=cut

my $msg = sub {
 "Indirect call of method \"$_[1]\" on object \"$_[0]\" at $_[2] line $_[3].\n"
};

sub unimport {
 shift;

 my $hook;
 while (@_) {
  my $arg = shift;
  if ($arg eq 'hook') {
   $hook = shift;
  } elsif ($arg eq ':fatal') {
   $hook = sub { die $msg->(@_) };
  }
  last if $hook;
 }
 $hook = sub { warn $msg->(@_) } unless defined $hook;

 $^H |= 0x00020000;
 $^H{+(__PACKAGE__)} = _tag($hook);

 ();
}

=head2 C<import>

Magically called at each C<use indirect>. Turns the module off.

=cut

sub import {
 $^H{+(__PACKAGE__)} = undef;
 ();
}

=head1 CONSTANTS

=head2 C<I_THREADSAFE>

True iff the module could have been built when thread-safety features.

=head1 CAVEATS

The implementation was tweaked to work around several limitations of vanilla C<perl> pragmas : it's thread safe, and doesn't suffer from a C<perl 5.8.x-5.10.0> bug that causes all pragmas to propagate into C<require>d scopes.

C<meth $obj> (no semicolon) at the end of a file won't be seen as an indirect object syntax, although it will as soon as there is another token before the end (as in C<meth $obj;> or C<meth $obj 1>).

With 5.8 perls, the pragma does not propagate into C<eval STRING>.
This is due to a shortcoming in the way perl handles the hints hash, which is addressed in perl 5.10.

=head1 DEPENDENCIES

L<perl> 5.8.

L<XSLoader> (standard since perl 5.006).

=head1 AUTHOR

Vincent Pit, C<< <perl at profvince.com> >>, L<http://www.profvince.com>.

You can contact me by mail or on C<irc.perl.org> (vincent).

=head1 BUGS

Please report any bugs or feature requests to C<bug-indirect at rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=indirect>.
I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc indirect

Tests code coverage report is available at L<http://www.profvince.com/perl/cover/indirect>.

=head1 ACKNOWLEDGEMENTS

Bram, for motivation and advices.

=head1 COPYRIGHT & LICENSE

Copyright 2008-2009 Vincent Pit, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut

1; # End of indirect
